src/v/storage/segment.cc

```diff
@@ -539,7 +539,7 @@ void segment::advance_stable_offset(size_t offset) {
 
     _reader.set_file_size(it->first);
     _tracker.stable_offset = it->second;
-    _inflight.erase(_inflight.begin(), it);
+    _inflight.erase(_inflight.begin(), std::next(it));
```

Comment:
> Fixed the problem by removing the iterator related with currently
processed in-flight update.

this looks correct, but despite the commit message claiming this is a problem, it doesn't seem problematic (e.g. it would have conservative semantics, presumably). perhaps stating why/what problem is fixed would be useful for others looking at git-blame in the future.

Reply:
Of course, i will include detailed description but briefly the problem is when truncation happens. the hard flush from segment appender should flush all the in flights but it does not as there is always that one which is not deleted. This one in-flight entry has physical offset larger than all subsequent ones because of truncation but the logical offset is smaller. This allow the reader to read past the stable offset.

Reply:
>This allow the reader to read past the stable offset.

this last part is unclear to me. if we are _not_ removing an inflight request which otherwise could be removed, then wouldn't the affect of that be that a reader couldn't read as would otherwise be safe to?

Reply:
i think the problem is a little bit more complex and the issue still exists however the EOF returned from reader input stream is handed differently. 

In a situation when we have batches exceeding single inflight write size there are multiple in flight operations for the same offset, but the first one already updates stable offset  <-- this is incorrect.

Reply:
this PR doesn't include the actual fix as it does not postpone updating the stable offset until last in flight operation pending for that offset finishes. I need to add this. 

Removing inflight operation changes a way how input stream is created and instead of hitting file EOF we hit its internal logical limit causing short read.

Reply:
very interesting, thanks. gonna look at the reproducer

Reply:
So i looked at this once again, the behavior looks correct, the EOF errors we saw are the normal behavior when reader reaches end of file.